tools/style.sh

```diff
@@ -9,7 +9,7 @@ fi
 Rscript -e "if (!require('styler')) install.packages('styler')"
 Rscript -e 'styler::style_pkg(exclude_files = list.files("./R", pattern = "^gen-*.*|^RcppExports.*", full.names = TRUE))'
 
-find . -type f \( -name 'DESCRIPTION' -o -name "*.R" \) ! -path "*/gen-*.*" ! -path "*/RcppExports.*" -exec sed -i -e 's/[ \t]*$//' {} \;
+find . -type f \( -name 'DESCRIPTION' -o -name "*.R" \) ! -path "*/gen-*.*" ! -path "*/RcppExports.*" -exec perl -pi -e 's/[ \t]*$//' {} \;
 
-find . -type f \( -name '*.h' -o -name '*.hpp' -o -name '*.c' -o -name '*.cc' -o -name '*.cpp' -o -name '*.cxx' \) ! -path "*/gen-*.*" ! -path "*/lantern.*"  ! -path "*/RcppExports.*" -exec sed -i -e 's/[ \t]*$//' {} \;
+find . -type f \( -name '*.h' -o -name '*.hpp' -o -name '*.c' -o -name '*.cc' -o -name '*.cpp' -o -name '*.cxx' \) ! -path "*/gen-*.*" ! -path "*/lantern.*"  ! -path "*/RcppExports.*" -exec perl -pi -e 's/[ \t]*$//' {} \;
```

Comment:
It would be nice to also exclude files in the `build` directories eg: `lantern/build` and `lantern/headers/build` as there might be plenty of files there in a common dev setup and they take a lot of time to format.

Actually, after formatting affects I can't compile lantern with `source("tools/buildlantern.R")`.
It seems some of the header files copied to `lantern/build` can be corrupted by the formating?

Reply:
I've already built `lantern` (+ `torch` and all of its dependencies) with the changes in this PR. It seems to me that there's a compatibility problem on your system. When I run `tools/style.sh` in this branch, I get no difference now.

Can you test this on Ubuntu? If not, I'll automate the style in GHA workflow.

The formatting takes a few seconds, after this PR (styling R code takes longer initially):
```bash
$ clang-format --version
clang-format version 10.0.0-4ubuntu1
$ time ./tools/style.sh
Loading required package: styler
Styling  189  files:
 R/autograd.R                                            ✔
 R/backends.R                                            ✔
 R/call_torch_function.R                                 ✔
 R/codegen-utils.R                                       ✔
 R/compat-purrr.R                                        ✔
 R/compilation_unit.R                                    ✔
 R/conditions.R                                          ✔
 R/contrib.R                                             ✔
 R/creation-ops.R                                        ✔
 R/cuda.R                                                ✔
 R/device.R                                              ✔
 R/dimname_list.R                                        ✔
 R/distributions-bernoulli.R                             ✔
 R/distributions-categorical.R                           ✔
 R/distributions-chi2.R                                  ✔
 R/distributions-constraints.R                           ✔
 R/distributions-exp-family.R                            ✔
 R/distributions-gamma.R                                 ✔
 R/distributions-mixture_same_family.R                   ✔
 R/distributions-multivariate_normal.R                   ✔
 R/distributions-normal.R                                ✔
 R/distributions-poisson.R                               ✔
 R/distributions-utils.R                                 ✔
 R/distributions.R                                       ✔
 R/dtype.R                                               ✔
 R/help.R                                                ✔
 R/indexing.R                                            ✔
 R/install.R                                             ✔
 R/ivalue.R                                              ✔
 R/jit-compile.R                                         ✔
 R/lantern_load.R                                        ✔
 R/lantern_sync.R                                        ✔
 R/layout.R                                              ✔
 R/linalg.R                                              ✔
 R/memory_format.R                                       ✔
 R/nn_adaptive.R                                         ✔
 R/nn-activation.R                                       ✔
 R/nn-batchnorm.R                                        ✔
 R/nn-conv.R                                             ✔
 R/nn-distance.R                                         ✔
 R/nn-dropout.R                                          ✔
 R/nn-init.R                                             ✔
 R/nn-linear.R                                           ✔
 R/nn-loss.R                                             ✔
 R/nn-normalization.R                                    ✔
 R/nn-pooling.R                                          ✔
 R/nn-rnn.R                                              ✔
 R/nn-sparse.R                                           ✔
 R/nn-upsampling.R                                       ✔
 R/nn-utils-clip-grad.R                                  ✔
 R/nn-utils-rnn.R                                        ✔
 R/nn-utils.R                                            ✔
 R/nn.R                                                  ✔
 R/nnf-activation.R                                      ✔
 R/nnf-batchnorm.R                                       ✔
 R/nnf-conv.R                                            ✔
 R/nnf-distance.R                                        ✔
 R/nnf-dropout.R                                         ✔
 R/nnf-embedding.R                                       ✔
 R/nnf-fold.R                                            ✔
 R/nnf-instancenorm.R                                    ✔
 R/nnf-linear.R                                          ✔
 R/nnf-loss.R                                            ✔
 R/nnf-normalization.R                                   ✔
 R/nnf-padding.R                                         ✔
 R/nnf-pixelshuffle.R                                    ✔
 R/nnf-pooling.R                                         ✔
 R/nnf-upsampling.R                                      ✔
 R/nnf-vision.R                                          ✔
 R/operators.R                                           ✔
 R/optim-adadelta.R                                      ✔
 R/optim-adagrad.R                                       ✔
 R/optim-adam.R                                          ✔
 R/optim-asgd.R                                          ✔
 R/optim-lbfgs.R                                         ✔
 R/optim-lr_scheduler.R                                  ✔
 R/optim-rmsprop.R                                       ✔
 R/optim-rprop.R                                         ✔
 R/optim-sgd.R                                           ✔
 R/optim.R                                               ✔
 R/package.R                                             ✔
 R/qscheme.R                                             ✔
 R/quantization.R                                        ✔
 R/R7.R                                                  ✔
 R/reduction.R                                           ✔
 R/save.R                                                ✔
 R/scalar.R                                              ✔
 R/script_module.R                                       ✔
 R/stack.R                                               ✔
 R/storage.R                                             ✔
 R/tensor_options.R                                      ✔
 R/tensor.R                                              ✔
 R/threads.R                                             ✔
 R/trace.R                                               ✔
 R/type-info.R                                           ✔
 R/utils-data-collate.R                                  ✔
 R/utils-data-dataloader.R                               ✔
 R/utils-data-enum.R                                     ✔
 R/utils-data-fetcher.R                                  ✔
 R/utils-data-sampler.R                                  ✔
 R/utils-data.R                                          ✔
 R/utils-pipe.R                                          ✔
 R/utils.R                                               ✔
 R/variable_list.R                                       ✔
 R/with-indices.R                                        ✔
 R/wrapers.R                                             ✔
 tests/testthat.R                                        ✔
 tests/testthat/helper-distributions.R                   ✔
 tests/testthat/helper-optim.R                           ✔
 tests/testthat/helper-tensor.R                          ✔
 tests/testthat/test-0-threads.R                         ✔
 tests/testthat/test-allocator.R                         ✔
 tests/testthat/test-autograd.R                          ✔
 tests/testthat/test-call_torch_function.R               ✔
 tests/testthat/test-contrib.R                           ✔
 tests/testthat/test-creation-ops.R                      ✔
 tests/testthat/test-cuda.R                              ✔
 tests/testthat/test-device.R                            ✔
 tests/testthat/test-dim-error-messages.R                ✔
 tests/testthat/test-dimname_list.R                      ✔
 tests/testthat/test-distribution-chi2.R                 ✔
 tests/testthat/test-distribution-gamma.R                ✔
 tests/testthat/test-distribution-normal.R               ✔
 tests/testthat/test-distribution-poisson.R              ✔
 tests/testthat/test-distributions-bernoulli.R           ✔
 tests/testthat/test-distributions-categorical.R         ✔
 tests/testthat/test-distributions-constraints.R         ✔
 tests/testthat/test-distributions-mixture_same_family.R ✔
 tests/testthat/test-distributions-multivariate_normal.R ✔
 tests/testthat/test-dtype.R                             ✔
 tests/testthat/test-gen-method.R                        ✔
 tests/testthat/test-gen-namespace.R                     ✔
 tests/testthat/test-generator.R                         ✔
 tests/testthat/test-indexing.R                          ✔
 tests/testthat/test-install.R                           ✔
 tests/testthat/test-ivalue.R                            ✔
 tests/testthat/test-jit-compile.R                       ✔
 tests/testthat/test-linalg.R                            ✔
 tests/testthat/test-memory_format.R                     ✔
 tests/testthat/test-nn-activation.R                     ✔
 tests/testthat/test-nn-batchnorm.R                      ✔
 tests/testthat/test-nn-conv.R                           ✔
 tests/testthat/test-nn-init.R                           ✔
 tests/testthat/test-nn-linear.R                         ✔
 tests/testthat/test-nn-loss.R                           ✔
 tests/testthat/test-nn-multihead-attention.R            ✔
 tests/testthat/test-nn-normalization.R                  ✔
 tests/testthat/test-nn-rnn.R                            ✔
 tests/testthat/test-nn-sparse.R                         ✔
 tests/testthat/test-nn-upsampling.R                     ✔
 tests/testthat/test-nn-utils-clip-grad.R                ✔
 tests/testthat/test-nn-utils-rnn.R                      ✔
 tests/testthat/test-nn.R                                ✔
 tests/testthat/test-nnf-embedding.R                     ✔
 tests/testthat/test-nnf-loss.R                          ✔
 tests/testthat/test-nnf-normalization.R                 ✔
 tests/testthat/test-nnf-padding.R                       ✔
 tests/testthat/test-nnf-pooling.R                       ✔
 tests/testthat/test-nnf-upsampling.R                    ✔
 tests/testthat/test-operators.R                         ✔
 tests/testthat/test-optim-adadelta.R                    ✔
 tests/testthat/test-optim-adagrad.R                     ✔
 tests/testthat/test-optim-adam.R                        ✔
 tests/testthat/test-optim-asgd.R                        ✔
 tests/testthat/test-optim-lbfgs.R                       ✔
 tests/testthat/test-optim-lr_scheduler.R                ✔
 tests/testthat/test-optim-rmsprop.R                     ✔
 tests/testthat/test-optim-rprop.R                       ✔
 tests/testthat/test-optim-sgd.R                         ✔
 tests/testthat/test-qscheme.R                           ✔
 tests/testthat/test-quantization.R                      ✔
 tests/testthat/test-R7.R                                ✔
 tests/testthat/test-readme.R                            ✔
 tests/testthat/test-reduction.R                         ✔
 tests/testthat/test-save.R                              ✔
 tests/testthat/test-scalar.R                            ✔
 tests/testthat/test-script_module.R                     ✔
 tests/testthat/test-stack.R                             ✔
 tests/testthat/test-storage.R                           ✔
 tests/testthat/test-tensor_options.R                    ✔
 tests/testthat/test-tensor.R                            ✔
 tests/testthat/test-trace.R                             ✔
 tests/testthat/test-utils-data-dataloader.R             ✔
 tests/testthat/test-utils-data-enum.R                   ✔
 tests/testthat/test-utils-data-sampler.R                ✔
 tests/testthat/test-utils-data.R                        ✔
 tests/testthat/test-variable_list.R                     ✔
 tests/testthat/test-with-indices.R                      ✔
 tests/testthat/test-wrapers.R                           ✔
───────────────────────────────────────────────────────
Status  Count   Legend
✔       189     File unchanged.
ℹ       0       File changed.
✖       0       Styling threw an error.
───────────────────────────────────────────────────────
Formatting ./inst/include/torch.h
Formatting ./inst/include/lantern/types.h
Formatting ./inst/include/utils.h
Formatting ./inst/include/torch_imports.h
Formatting ./inst/include/torch_impl.h
Formatting ./inst/include/torch_api.h
Formatting ./inst/include/torch_types.h
Formatting ./inst/include/torch_deleters.h
Formatting ./lantern/headers/src/main.cpp
Formatting ./lantern/include/lantern/types.h
Formatting ./lantern/src/JITTypes.cpp
Formatting ./lantern/src/TensorOptions.cpp
Formatting ./lantern/src/Stack.cpp
Formatting ./lantern/src/base64.hpp
Formatting ./lantern/src/utils.cpp
Formatting ./lantern/src/MemoryFormat.cpp
Formatting ./lantern/src/Device.cpp
Formatting ./lantern/src/NNUtilsRnn.cpp
Formatting ./lantern/src/TensorList.cpp
Formatting ./lantern/src/Tensor.cpp
Formatting ./lantern/src/ScriptModule.cpp
Formatting ./lantern/src/Allocator.cpp
Formatting ./lantern/src/Scalar.cpp
Formatting ./lantern/src/Function.h
Formatting ./lantern/src/Contrib/SortVertices/utils.h
Formatting ./lantern/src/Contrib/SortVertices/sort_vert.cpp
Formatting ./lantern/src/Contrib/SortVertices/sort_vert.h
Formatting ./lantern/src/Contrib/SortVertices/sort_vert_cpu.cpp
Formatting ./lantern/src/Contrib/SortVertices/cuda_utils.h
Formatting ./lantern/src/Contrib/Sparsemax.cpp
Formatting ./lantern/src/Threads.cpp
Formatting ./lantern/src/Layout.cpp
Formatting ./lantern/src/Dimname.cpp
Formatting ./lantern/src/Compile.cpp
Formatting ./lantern/src/Autograd.cpp
Formatting ./lantern/src/Dtype.cpp
Formatting ./lantern/src/Delete.cpp
Formatting ./lantern/src/Storage.cpp
Formatting ./lantern/src/Indexing.cpp
Formatting ./lantern/src/Trace.cpp
Formatting ./lantern/src/utils.hpp
Formatting ./lantern/src/Reduction.cpp
Formatting ./lantern/src/Generator.cpp
Formatting ./lantern/src/Backends.cpp
Formatting ./lantern/src/AllocatorCuda.cpp
Formatting ./lantern/src/Cuda.cpp
Formatting ./lantern/src/Function.cpp
Formatting ./lantern/src/Quantization.cpp
Formatting ./lantern/src/AllocatorUtils.h
Formatting ./lantern/src/QScheme.cpp
Formatting ./lantern/src/IValue.cpp
Formatting ./lantern/src/Save.cpp
Formatting ./lantern/tests/init.h
Formatting ./lantern/tests/init.cpp
Formatting ./lantern/tests/main.cpp
Formatting ./src/codegen.cpp
Formatting ./src/backends.cpp
Formatting ./src/scalar.cpp
Formatting ./src/utils.cpp
Formatting ./src/device.cpp
Formatting ./src/autograd.cpp
Formatting ./src/tensor_options.cpp
Formatting ./src/cuda.cpp
Formatting ./src/script_module.cpp
Formatting ./src/reduction.cpp
Formatting ./src/trace.cpp
Formatting ./src/contrib.cpp
Formatting ./src/utils.h
Formatting ./src/torch_api.cpp
Formatting ./src/layout.cpp
Formatting ./src/tensor.cpp
Formatting ./src/torch_exports.cpp
Formatting ./src/variable_list.cpp
Formatting ./src/allocator.cpp
Formatting ./src/generator.cpp
Formatting ./src/dimname_list.cpp
Formatting ./src/stack.cpp
Formatting ./src/memory_format.cpp
Formatting ./src/storage.cpp
Formatting ./src/ivalue.cpp
Formatting ./src/xptr.cpp
Formatting ./src/quantization.cpp
Formatting ./src/tensor_list.cpp
Formatting ./src/nn_utils_rnn.cpp
Formatting ./src/indexing.cpp
Formatting ./src/jit-compile.cpp
Formatting ./src/translate_messages.h
Formatting ./src/save.cpp
Formatting ./src/qscheme.cpp
Formatting ./src/dtype.cpp

real    0m10.756s
user    0m4.026s
sys     0m6.283s
```
```

Reply:
Also, after styling/formatting, we need to run `roxygen2::roxygenize()`, which is I did here.

Reply:
@dfalbel The differences you report could be whitespaces; can you confirm? If so, 30cb3122a57936d8e1a3a546957ee1a234aaae54 should fix it.

As for the processing time, I can't reproduce your issue. I'll test building `lantern` again with the R script.

Reply:
Yeah, I think they could be related to whitespaces, let me re-check!
Also, I merged #755 that was somewhat large PR. Sorry, for making extra work, but I think it's easier to rebase this PR than the other one.

Reply:
No worries. I've merged the `master` branch with restyling in f5e0b9ec5cfc1d77d61211123fc9cfcc8f75749f, which makes the changes in #755 cleaner (no whitespaces before and after).

Regarding `lantern` build, doesn't we already test it in the main GHA workflow? If so, it passes that test; if not, we should really cover this test in GHA.

Reply:
I see. After 8ea63f834148963d746ef8fd7f63382d1c5e6b2f (and any required fixes/updates), we wont' need to run styling locally. The GHA workflow will style and render documents (+ commit changes) if R-CMD-check is successful (on Linux + R-release; I'd like to R-devel, you probably have a reason to exclude it from the matrix).

Reply:
The CI style worked (just failed due to a typo). From the log, it seems that the differences you found earlier in the R code are due to `styler` version (I use development). Anyway, GHA will automatically fix this now.